README.md
<div align="center">
<p align="center">
  <img alt="logo" src="https://raw.githubusercontent.com/mbarkhau/lib3to6/master/lib3to6_128.png">
</p>
</div>

# [lib3to6][repo_ref]

Compile Python 3.7+ code to Python 2.7+ compatible code. The idea is
quite similar to Babel https://babeljs.io/. Develop using the newest
interpreter and use (most) new language features and still maintain
backward compatibility.

Project/Repo:

[![MIT License][license_img]][license_ref]
[![Supported Python Versions][pyversions_img]][pyversions_ref]
[![CalVer v202110.1050-b0][version_img]][version_ref]
[![PyPI Version][pypi_img]][pypi_ref]
[![PyPI Downloads][downloads_img]][downloads_ref]

Code Quality/CI:

[![GitHub CI Status][github_build_img]][github_build_ref]
[![GitLab CI Status][gitlab_build_img]][gitlab_build_ref]
[![Type Checked with mypy][mypy_img]][mypy_ref]
[![Code Coverage][codecov_img]][codecov_ref]
[![Code Style: sjfmt][style_img]][style_ref]


|               Name                  |    role           |  since  | until |
|-------------------------------------|-------------------|---------|-------|
| Manuel Barkhau (mbarkhau@gmail.com) | author/maintainer | 2018-09 | -     |


<!--
  To update the TOC:
  $ pip install md-toc
  $ md_toc -i gitlab README.md -l 2
-->


[](TOC)

- [Motivation: Ease the Transition from Old Interpreters](#motivation-ease-the-transition-from-old-interpreters)
- [Python Versions and Compatibility](#python-versions-and-compatibility)
- [Usage Caveats](#usage-caveats)
- [Integration with `setup.py`](#integration-with-setuppy)
- [Deprecations](#deprecations)
- [Per-File Opt-In/Opt-Out](#per-file-opt-inopt-out)
- [Automatic Conversions](#automatic-conversions)
- [Motivation](#motivation)
- [How it works](#how-it-works)
- [Contributing](#contributing)
- [Project Status (as of 2021-01-01): Beta](#project-status-as-of-2021-01-01-beta)
- [Future Work](#future-work)
- [Alternatives](#alternatives)
- [FAQ](#faq)

[](TOC)


## Motivation: Ease the Transition from Old Interpreters

Especially for packages, you probably don't want to force all your
users to already use python 3.10, so maintaining backward
compatibility is a good idea. Nonetheless, for development it is a
good idea to use the newest interpreter you can and ideally you can
use the newer type annotation features to type-check your code, but
ellide these annotations in the distribution, since they are not
needed at runtime.

If your existing project uses Python2.7, it may not be possible to
dedicate a substantial block of time to update all of your code and
then flip a switch to start running on Python3. To make matters worse,
you may continue to write code only for Python2.7, since that is what
your production code will actually run on. With `lib3to6` you can
start to use Python3 for development and integration (ensuring forward
compatibility) and still maintain backward compatibility while you
have to deploy for Python2.


## Python Versions and Compatibility

The compiled output is tested using:

- Python 3.10
- Python 3.9
- Python 3.8
- Python 3.7
- Python 2.7
- PyPy 3.7

The test-suite for the transpiler is run using:

- Python 3.10
- Python 3.9
- Python 3.8
- Python 3.7
- PyPy 3.7

The compiled output may work with other versions of python, such as `<=2.6` or `>=3.0 <=3.6`, but these are not tested.


## Usage Caveats

`lib3to6` does not add any runtime dependencies of its own, but it does inject code, such as temporary variables and imports from the standard library (`itertools` and `builtins` in particular). Any changes will only add a constant `O(1)` overhead.

`lib3to6` does optimistic ast transformations, with the assumption that you're not doing anything too crazy in your code. An example of such a transformation is the support for [PEP3102 - Keyword-Only Arguments](https://www.python.org/dev/peps/pep-3102/). `lib3to6` will change the function signature to use `**kwargs` and add locals extracted from `kwargs`.

```
$ cat kwonly_args_demo.py
def compare(a, b, *, key=None):
    ...
```

```
$ lib3to6 kwonly_args_demo.py
def compare(a, b, **kwargs):
    key = kwargs.get('key', None)
    pass
```

This means that the function signature you can get using the `inspect` module may not be what you expect for the output of `lib3to6`.


## Integration with `setup.py`

The cli command `lib3to6 <filename>` is nice for demo purposes, but
for integration with your project, you may prefer to use it in your
`setup.py` file. Contributions for other kinds of integration are most
welcome.


```python
# setup.py
try:
    import lib3to6
    distclass = lib3to6.Distribution
except ImportError:
    distclass = setuptools.dist.Distribution

setuptools.setup(
    name="my-module",
    license="MIT",
    ...
    python_requires=">=2.7",
    install_requires=[...],
    distclass=distclass,
)
```

When you build you package, the contents of the resulting distribution
will be the code that was converted by lib3to6.


```bash
~/my-module $ python setup.py bdist_wheel --python-tag=py2.py3
running bdist_wheel
...
~/my-module$ ls -1 dist/
my_module-201808.1-py2.py3-none-any.whl

~/my-module$ python3 -m pip install dist/my_module-201808.1-py2.py3-none-any.whl
Processing ./dist/my_module-201808.1-py2.py3-none-any.whl
Installing collected packages: my-module
Successfully installed my-module-201808.1

~/my-module$ python2 -m pip install dist/my_module-201808.1-py2.py3-none-any.whl
Processing ./dist/my_module-201808.1-py2.py3-none-any.whl
Installing collected packages: my-module
Successfully installed my-module-201808.1
```

When testing, make sure you're not importing `my_module` from your local
directory, which is probably the original source code. Instead you can
either manipulate your `PYTHONPATH`, or simply switch directories...

```bash
~/$ python3 -c "import my_module"
/home/user/my-module/my_module/__init__.py
Hello 世界 from 3.6.5!

~/my-module$ cd ..
~/$ python3 -c "import my_module"
/home/user/envs/py36/lib/python3.6/site-packages/my_module/__init__.py
Hello 世界 from 3.6.5!

~$ python2 -c "import my_module"
/home/user/envs/py27/lib/python2.7/site-packages/my_module/__init__.py
Hello 世界 from 2.7.15!
```


## Deprecations

Before `v202108.1048` the recomendation was to use
`package_dir = lib3to6.fix(package_dir, ...)`.
This approach did not make use of metadata already provided to
`setuptools.setup()` and also caused issues when using
`include_package_data=True`. Use the above described
`lib3to6.Distribution` instead.


## Per-File Opt-In/Opt-Out

Since `lib3to6==v202009.1044` there is support to selectively
enable/disable transpilation on a per-file basis.

Any file which starts with a `# lib3to6: disabled` comment, will not
be transpiled. For these, you will have to take care of
forward/backward compatibility yourself.

```python
# -*- coding: utf-8 -*-
# lib3to6: disabled
"""A module written to work both with Python2 and 3.

This module is not transpiled by lib3to6.
"""

from __future__ import print_function
...

import sys

PY3 = sys.version_info[0] > 2

if PY3:
    ...
else:
    ...
```

Instead of opt-out, you can also take an opt-in approach. You will
have to use the `lib3to6_default_mode` option:

```python
# setup.py
try:
    import lib3to6
    distclass = lib3to6.Distribution
except ImportError:
    distclass = setuptools.dist.Distribution


setuptools.setup(
    ...
    distclass=distclass,
    lib3to6_default_mode='disabled',     # default: enabled
)
```

This will leave all files untouched, except for those marked with a `# lib3to6: enabled` comment.

```python
# lib3to6: enabled
"""A module written to work both with Python2 and 3.

This module is transpiled with lib3to6.
"""

name: str = "Wörld"
print(f"Hello {world}!")
```


## Automatic Conversions

Not all new language features have a semantic equivalent in older
versions. To the extent these can be detected, an error will be
reported when these features are used.

Note that a fix is not applied if the lowest version of python that
you are targeting already supports the newer syntax. The conversions
are ordered by when the feature was introduced.


### PEP 572: Assignment Expressions (aka. the walrus operator)

```python
# Since 3.8
if match1 := pattern1.match(data):
    result = match1.group(1)

# From 2.7 to 3.7
match1 = pattern1.match(data)
if match1:
    result = match1.group(1)
```

Some expressions nested expressions in a condition are not so easy,
in which case lib3to6 will bend over backwards.

```python
# Since 3.8
while (block := f.read(4096)) != '':
    process(block)

# For [2.7 - 3.7]
__loop_condition = True
while __loop_condition:
    block = f.read(4096)
    __loop_condition = block != ''
    if __loop_condition:
        process(block)
```


### PEP 563: Postponed Evaluation of Annotations

```python
# Since 3.7
class SelfRef:
    def method(self) -> SelfRef:
        pass

# From 3.0 to 3.6
class SelfRef:
    def method(self) -> 'SelfRef':
        pass
```

Note that this is not a stupid conversion that is applied to all
annotations, it is only applied to annotations that are forward
references. Backward references are left as is.


```python
# Since 3.7
class BackRef:
    def method(self) -> ForwardRef:
        pass

class ForwardRef:
    def method(self) -> BackRef:
        pass


# From 3.0 to 3.6
class BackRef:
    def method(self) -> 'ForwardRef':
        pass

class ForwardRef:
    def method(self) -> BackRef:
        pass
```

If you're supporting python 2.7, the annotation will of course be elided.


### PEP 498: formatted string literals.

```python
# Since 3.6
who = "World"
print(f"Hello {who}!")

# From 2.7 to 3.5
who = "World"
print("Hello {0}!".format(who))
```

The fixer also converts the newer `{var=}` syntax, even if you use
lib3to6 on a Python version older than 3.8.

```python
# Since 3.6
who = "World"
print(f"Hello {who=}!")

# From 2.7 to 3.5
print("Hello who={0}!".format(who))
```


### Eliding of Annotations

```python
# Since 3.0
def foo(bar: int) -> str:
    pass

# In 2.7
def foo(bar):
    pass
```


### PEP 515: underscores in numeric literals

```python
# Since 3.6
num = 1_234_567

# From 2.7 to 3.5
num = 1234567
```


### Unpacking generalizations

For literals...

```python
# Since 3.4
x = [*[1, 2], 3]

# From 2.7 to 3.3
x = [1, 2, 3]
```

For varargs...

```python
# Since 3.4
foo(0, *a, *b)

# From 2.7 to 3.3
foo(*([0] + list(a) + list(b))
```

For kwargs...

```python
# Since 3.4
foo(**x, y=22, **z)

# From 2.7 to 3.3
import itertools
foo(**dict(itertools.chain(x.items(), {'y': 22}.items(), z.items())))
```

Note that the import will only be added to your module once.


### Keyword only arguments


```python
# Since 3.6
def kwonly_func(*, kwonly_arg=1):
    ...

# From 2.7 to 3.5
def kwonly_func(**kwargs):
    kwonly_arg = kwargs.get('kwonly_arg', 1)
    ...
```


### Convert class based typing.NamedTuple usage to assignments


```python
import typing

# Since 3.5
class Bar(typing.NamedTuple):
    x: int
    y: str

# From 2.7 to 3.4
Bar = typing.NamedTuple('Bar', [('x', int), ('y', str)])
```


### New Style Classes

```python
# Since 3.0
class Bar:
  pass

# Before 3.0
class Bar(object):
  pass
```


### Future Imports

All `__future__` imports applicable to your target version are
prepended to every file.

```python
# -*- coding: utf-8 -*-
# This file is part of the <X> project
# ...
"""A docstring."""

x = True
```

With `target-version=27` (the default).

```python
# -*- coding: utf-8 -*-
# This file is part of the <X> project
# ...
"""A docstring."""
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

x = True
```

With `target-version=3.7`

```python
# -*- coding: utf-8 -*-
# This file is part of the <X> project
# ...
"""A docstring."""
from __future__ import annotations

x = True
```

Note that `lib3to6` works mostly at the ast level, but an exception is
made for any comments that appear at the top of the file. These are
preserved as is, so your shebang, file encoding and licensing headers
will be preserved.


### Not Supported Features

An (obviously non-exhaustive) list of features which are **not
supported**, either because they involve a semantic change, or
because there is no simple ast transformation to make them work
across different python versions:

 - PEP 492 - `async`/`await`
 - PEP 465 - `@`/`__matmul__` operator
 - PEP 380 - `yield from` syntax
 - PEP 584 - union operators for `dict`
 - ordered dictionary (since python 3.6)


### Modules with Backports

Some new modules have backports, which lib3to6 will point to:

 - typing
 - pathlib -> pathlib2
 - secrets -> python2-secrets
 - ipaddress -> py2-ipaddress
 - csv -> backports.csv
 - lzma -> backports.lzma
 - enum -> enum34

For a full list of modules for which these warnings and errors apply,
please review [`MAYBE_UNUSABLE_MODULES` in
src/lib3to6/checkers_backports.py](https://gitlab.com/mbarkhau/lib3to6/blob/master/src/lib3to6/checkers_backports.py)

For some modules, the backport uses the same module name as the original
module in the standard library. By default, lib3to6 will only warn about
usage of such modules, since it cannot detect if you're using the module
from the backported package (good) or from the standard library (bad if
not available in your target version). If you would like to opt-in to hard
error messages, you can whitelist modules for which you have the
backported package as a dependency.

A good approach to adding such backports as dependencies is to
qualify the requirement with a [dependency
specification](https://www.python.org/dev/peps/pep-0508/), so that
users with a newer interpreter use the builtin module and don't
install the backport package that they don't need.

These work as arguments for `install_requires` and also in
`requirements.txt` files.

```python
import setuptools

setuptools.setup(
    name="my-package",
    install_requires=['typing;python_version<"3.5"'],
    ...
)
```

For testing, you can also pass these as a space separated parameter
to the `lib3to6` cli command:

```shell
$ lib3to6 my_script.py > /dev/null
WARNING - my_script.py@1: Use of import 'enum'.
    This module is only available since Python 3.5,
    but you configured target_version=2.7.
WARNING - my_script.py@2: Use of import 'typing'.
    This module is only available since Python 3.5,
    but you configured target_version=2.7.

import enum
import typing
...

$ lib3to6 `--install-requires='typing'` my_script.py > /dev/null
Traceback (most recent call last):
  ...
  File "/home/user/.../lib3to6/src/lib3to6/checkers_backports.py", line 134, in __call__
    raise common.CheckError(errmsg, node)
lib3to6.common.CheckError: my_script.py@1 - Prohibited import 'enum'.
    This module is available since Python 3.4,
    but you configured target_version='2.7'.
    Use 'https://pypi.org/project/enum34' instead.

$ lib3to6 `--install-requires='typing enum34'` my_script.py
import enum
import typing
...
```


## Motivation

The main motivation for this project is to be able to use `mypy`
without sacrificing compatibility to older versions of python.

```python
# my_module/__init__.py
def hello(who: str) -> None:
    import sys
    print(f"Hello {who} from {sys.version.split()[0]}!")


print(__file__)
hello("世界")
```


```bash
$ pip install lib3to6
$ python -m lib3to6 my_module/__init__.py
```


```python
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals


def hello(who):
    import sys
    print('Hello {0} from {1}!'.format(who, sys.version.split()[0]))


print(__file__)
hello('世界')
```


Fixes are applied to match the semantics of python3 code as
close as possible, even when running on a python2.7 interpreter.

Some fixes that have been applied in the above:

    - PEP263 magic comment to declare the coding of the python
      source file. This allows the string literal `"世界"` to
      be decoded correctly.
    - `__future__` imports have been added. This includes the well
      known print statement -> function change. The unicode_literals
    - Type annotations have been removed
    - `f""` string -> `"".format()` conversion


### Compatibility Matters

I've seen a common gut reaction to lib3to6, which is that we shouldn't care about older versions of Python, Python 2.7 in particular. I would humbly suggest you consider the position of people other than developers who have full control over their development environment and only use CPython. As of this writing (August 2020), the most recent language version supported by alternative interpreters is the following:

| Interpreter | Version |
|-------------|---------|
| Stackless   |     3.7 |
| PyPy        |     3.6 |
| MicroPython |     3.4 |
| IronPython  |     2.7 |
| Jython      |     2.7 |

Notice that even Stackless Python, which has the least effort to keep up with new language features, is nonetheless lagging behind CPython. And even if all you care about is CPython, be aware that the most recent interpreter may not be available on platforms that users care about. For example on PythonAnywhere.com, the most recent version of CPython is 3.5 and of PyPy is 2.7.

If you are writing a library and it doesn't **need** any of the newer **runtime** features, such as async/await or ordered dictionaries, then I would humbly suggest you do not unnecessarily prevent users of such platforms from using your library.

From a users perspective, only supporting the newest versions of Python might be interpreted as arrogance, but your time as a maintainer isn't free, and you don't owe users of your library anything. Lib3to6 exists to minimize your effort to maintain backward compatibility. If you have difficulties integrating lib3to6 into your packaging process, please report an issue: [gitlab.com/mbarkhau/lib3to6/-/issues](https://gitlab.com/mbarkhau/lib3to6/-/issues).


### On Testing your Project

Projects that use lib3to6 should have a test-suite that is executed
with the oldest python version that you want to support, using the
converted output generated by lib3to6. While you can develop using a
newer version of python, you should not blindly trust lib3to6 as it is
very easy to introduce backward incompatible changes if you only test
on the most recent interpreter. The most obvious example is that
lib3to6 cannot do much to help you if a library produces `bytes` on
Python 2 but `str` on Python 3.

The easiest way I have found to test a project, is to create a
distribution using `python setup.py bdist_wheel` with the above
modifications to the `setup.py`, install it and run the test-suite
against the installed modules.


## How it works

This project works at the level of the python abstract syntax
tree (AST). The AST is transformed so that is only uses
constructs that are also valid in older versions of python. For
example it will translate f-strings to normal strings using the
``str.format`` method.

```python
>>> import sys
>>> sys.version_info
'3.6.5'
>>> import lib3to6
>>> py3_source = 'f"Hello {1 + 1}!"'
>>> cfg = {"fixers": ["f_string_to_str_format"]}
>>> py2_source = lib3to6.transpile_module(cfg, py3_source)

>>> print(py3_source)
f"Hello {1 + 1}!"
>>> print(py2_source)
# -*- coding: utf-8 -*-
"Hello {0}!".format(1 + 1)
```

At a lower level, this translation is based on detection of the
`ast.JoinedStr` node, which is translated into and AST that can be
serialized back into python syntax that will also work on older
versions.

```python
>>> print(lib3to6.parsedump_ast(py3_source))
Module(body=[Expr(value=JoinedStr(values=[
    Str(s='Hello '),
    FormattedValue(
        value=BinOp(left=Num(n=1), op=Add(), right=Num(n=1)),
        conversion=-1,
        format_spec=None,
    ),
    Str(s='!'),
]))])
>>> print(lib3to6.parsedump_ast(py2_source))
Module(body=[Expr(value=Call(
    func=Attribute(
        value=Str(s='Hello {0}!'),
        attr='format',
        ctx=Load(),
    ),
    args=[BinOp(left=Num(n=1), op=Add(), right=Num(n=1))],
    keywords=[]
))])
```

### Checker Errors

Of course this does not cover every aspect of compatibility.
Changes in APIs cannot be translated automatically in this way.

An obvious example, is that there is no way to transpile code
which uses `async` and `await`. In this case, `lib3to6`
will simply raise a CheckError. This applies only to your source
code though, so if import use a library which uses `async` and
`await`, everything may look fine until you run your tests
on python 2.7.

A more subtle example is the change in semantics of the builtin
`open` function.

```bash
$ cat open_example.py
with open("myfile.txt", mode="w", encoding="utf-8") as fobj:
    fobj.write("Hello Wörld!")
$ python2 open_example.py
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: 'encoding' is an invalid keyword argument for this function
```


Usually there are alternative ways to write equivalent code that
works on all versions of python. For these common
incompatibilities lib3to6 will raise an error and suggest an
alternative, such as in this case using `io.open` instead.

```bash
$ lib3to6 open_example.py
Traceback (Most recent call last):
11  lib3to6      <module>         --> sys.exit(main())
764 core.py      __call__         --> return self.main(*args, **kwargs)
717 core.py      main             --> rv = self.invoke(ctx)
956 core.py      invoke           --> return ctx.invoke(self.callback, **ctx.params)
555 core.py      invoke           --> return callback(*args, **kwargs)
55  __main__.py  main             --> fixed_source_text = transpile.transpile_module(cfg, source_text)
260 transpile.py transpile_module --> checker(cfg, module_tree)
158 checkers.py  __call__         --> raise common.CheckError(msg, node)
CheckError: Prohibited keyword argument 'encoding' to builtin.open. on line 1 of open_example.py
```


Here `lib3to6` you will give you a `CheckError`, however it
remains your responsibility to write your code so that this
syntactic translation is semantically equivalent in both python3
and python2.

`lib3to6` uses the python `ast` module to parse your code. This
means that you need a modern python interpreter to transpile from
modern python to legacy python interpreter. You cannot transpile
features which your interpreter cannot parse. The intended use is
for developers of libraries who use the most modern python
version, but want their libraries to work on older versions.


## Contributing

The most basic contribution you can make is to provide minimal,
reproducible examples of code that should either be converted or
which should raise an error.

The project is hosted at
[gitlab.com/mbarkhau/lib3to6](https://gitlab.com/mbarkhau/lib3to6),
mainly because that's where the CI/CD is configured. GitHub is only
used as a copy/backup (and because that seems to be where many people
look for things).

You can get started with local development in just a few commands.

```shell
user@host:~/ $ git clone https://gitlab.com/mbarkhau/lib3to6.git
user@host:~/ $ cd lib3to6/
user@host:~/lib3to6/ ⎇master $ make help
user@host:~/lib3to6/ ⎇master $ make conda     # creates conda environments
...
user@host:~/lib3to6/ ⎇master $ ls ~/miniconda3/envs/
user@host:~/lib3to6_pypy35 lib3to6_py27 lib3to6_py36 lib3to6_py37 lib3to6_py38
```

The targets in the makefile are set up to use the virtual environments.

```shell
user@host:~/lib3to6/ ⎇master $ make fmt
All done! ✨ 🍰 ✨
21 files left unchanged.

user@host:~/lib3to6/ ⎇master $ make lint mypy devtest
isort ... ok
sjfmt ... ok
flake8 .. ok
mypy .... ok
...
```


CHANGELOG.md
# Changelog for https://github.com/mbarkhau/lib3to6


## Next

 - Add fixer for type hinting generics `'x: list[int]` -> `x: typing.List[int]`
 - Add fixer for Union Operator `'x: A | B` -> `x: typing.Union[A, B]`


## v202110.1050

 - Add new default integration method: [`lib3to6.Distribution`][href_readme_integration]
 - Deprecate old integration method (had issues with package data).
 - Fix invalid parsing of install_requires.
 - Fix cache invalidation when build config changes.
 - Fix type annotations with constants (e.g. `Optional['MyClass']`)
 - Fix compat issues with python 3.10
 - Disable testing for Python 3.5 and Python 3.6


[href_readme_integration]: https://github.com/mbarkhau/lib3to6#integration-with-setuppy


## v202107.1047

 - Fix type annotations with list arguments


## v202107.1046

 - Fix type annotations with attribute access


## v202101.1045

 - Fixes for Python 3.9 and mypy 0.800


## v202009.1042

 - New #6: per-file opt-in/opt-out using `# lib3to6: disabled`/`# lib3to6: enabled`


## v202006.1041

 - New: Lots more documentation.
 - New #5: Add detection of invalid imports and point to available backports. Use `install_requires` option to whitelist installed backports.
 - New: Checkers produce better error messages.
 - New: Colouring of diffs when using `lib3to6` cli command.
 - New: Checker for `yield from` syntax on target version doesn't support it
 - New: Checker for `@` operator when target version doesn't support it
 - Fix #3: `--target-version` argument is ignored [gitlab../issues/3](https://gitlab.com/mbarkhau/lib3to6/-/issues/3)
 - Fix #4: Remove `from __future__ import X` when the target version doesn't support it.
 - Fix #4: Convert Forward Reference Annotations to strings [gitlab../issues/4](https://gitlab.com/mbarkhau/lib3to6/-/issues/4) Thank you [Faidon Liambotis](https://gitlab.com/paravoid) for your help with testing and helping to debug ❤️.
 - Fix: Don't apply keyword only args fixer for `--target-version=3.0` or above.


## v202002.0031

 - Compatibility fixes for Python 3.8
 - Add support for f-string `=` specifier
 - Add support for `:=` walrus operator (except inside comprehensions)


## v201902.0030

 - Fix python 2 builtins were not always overridden correctly.
 - Fix pypy compatibility testing
 - Better mypy coverage


## v201812.0021-beta

 - Recursively apply some fixers.


## v201812.0020-alpha

 - Move to gitlab.com
 - Use bootstrapit
 - Fix bugs based on use with pycalver


## v201809.0019-alpha

 - CheckErrors include log line numbers
 - Transpile errors now include filenames
 - Added fixers for renamed modules, e.g.
    .. code-block:: diff

        - import queue
        + try:
        +     import queue
        + except ImportError:
        +     import Queue as queue


## v201808.0014-alpha

 - Better handling of package_dir
 - Change to `CalVer Versioning <https://calver.org/>`_
 - Remove console script in favour of simple ``python -m lib3to6``
 - Rename from ``three2six`` -> ``lib3to6``


src/lib3to6/common.py
# This file is part of the lib3to6 project
# https://github.com/mbarkhau/lib3to6
#
# Copyright (c) 2019-2021 Manuel Barkhau (mbarkhau@gmail.com) - MIT License
# SPDX-License-Identifier: MIT

import ast
import typing as typ
import builtins

PackageName      = str
PackageDirectory = str
PackageDir       = typ.Dict[PackageName, PackageDirectory]

InstallRequires = typ.Optional[typ.Set[str]]


ConstantNodeTypes: typ.Tuple[typ.Type[ast.Constant], ...] = (ast.Constant,)

# Deprecated since version 3.8: Class ast.Constant is now used for all
# constants. Old classes ast.Num, ast.Str, ast.Bytes, ast.NameConstant and
# ast.Ellipsis are still available, but they will be removed in future Python
# releases.

if hasattr(ast, 'Num'):
    ConstantNodeTypes += (ast.Num, ast.Str, ast.Bytes, ast.NameConstant, ast.Ellipsis)


LeafNodeTypes = ConstantNodeTypes + (
    ast.Name,
    ast.cmpop,
    ast.boolop,
    ast.operator,
    ast.unaryop,
    ast.expr_context,
)


ContainerNodes = (ast.List, ast.Set, ast.Tuple)


class BuildConfig(typ.NamedTuple):

    target_version  : str  # e.g. "2.7"
    cache_enabled   : bool
    default_mode    : str
    fixers          : str
    checkers        : str
    install_requires: InstallRequires


class BuildContext(typ.NamedTuple):

    cfg     : BuildConfig
    filepath: str


def init_build_context(
    target_version  : str             = "2.7",
    cache_enabled   : bool            = True,
    default_mode    : str             = 'enabled',
    fixers          : str             = "",
    checkers        : str             = "",
    install_requires: InstallRequires = None,
    filepath        : str             = "<filepath>",
) -> BuildContext:
    cfg = BuildConfig(
        target_version=target_version,
        cache_enabled=cache_enabled,
        default_mode=default_mode,
        fixers=fixers,
        checkers=checkers,
        install_requires=install_requires,
    )
    return BuildContext(cfg=cfg, filepath=filepath)


# Additional items:
#   'filepath': "/path/to/inputfile.py"


class InvalidPackage(Exception):
    pass


def get_node_lineno(
    node: typ.Optional[ast.AST] = None, parent: typ.Optional[ast.AST] = None
) -> int:
    if isinstance(node, (ast.stmt, ast.expr)):
        return node.lineno
    elif isinstance(parent, (ast.stmt, ast.expr)):
        return parent.lineno
    else:
        return -1


class CheckError(Exception):

    lineno: int

    def __init__(self, msg: str, node: ast.AST = None, parent: ast.AST = None) -> None:
        self.lineno = get_node_lineno(node, parent)
        super().__init__(msg)


class FixerError(Exception):

    msg     : str
    node    : ast.AST
    parent  : typ.Optional[ast.AST]
    filepath: typ.Optional[str]

    def __init__(
        self,
        msg     : str,
        node    : ast.AST,
        parent  : typ.Optional[ast.AST] = None,
        filepath: typ.Optional[str    ] = None,
    ) -> None:
        self.msg      = msg
        self.node     = node
        self.parent   = parent
        self.filepath = filepath
        super().__init__(msg)

    def __str__(self) -> str:
        msg = self.msg
        if self.filepath:
            lineno = get_node_lineno(self.node, self.parent)
            msg += f" File: {self.filepath}@{lineno}"
        return msg


class VersionInfo:
    """Compatibility info for Fixer and Checker classes.

    Used as a class attribute (not an instance attribute) of a fixer or checker.
    The compatability info relates to the fixer/checker rather than the feature
    it deals with. The question that is being ansered is: "should this
    fixer/checker be exectued for this (src_version, tgt_version) pair"
    """

    # since/until is inclusive
    apply_since: typ.List[int]
    apply_until: typ.Optional[typ.List[int]]
    works_since: typ.List[int]
    works_until: typ.Optional[typ.List[int]]

    def __init__(
        self,
        apply_since: str = "1.0",
        apply_until: typ.Optional[str] = None,
        works_since: typ.Optional[str] = None,
        works_until: typ.Optional[str] = None,
    ) -> None:

        self.apply_since = [int(part) for part in apply_since.split(".")]
        if apply_until is None:
            self.apply_until = None
        else:
            self.apply_until = [int(part) for part in apply_until.split(".")]

        if works_since is None:
            # Implicitly, if it's applied since a version, it
            # also works since then.
            self.works_since = self.apply_since
        else:
            self.works_since = [int(part) for part in works_since.split(".")]

        if works_until is None:
            self.works_until = None
        else:
            self.works_until = [int(part) for part in works_until.split(".")]

    def is_required_for(self, version: str) -> bool:
        version_num = [int(part) for part in version.split(".")]
        apply_until = self.apply_until
        if apply_until and apply_until < version_num:
            return False
        else:
            return self.apply_since <= version_num

    def is_compatible_with(self, version: str) -> bool:
        version_num = [int(part) for part in version.split(".")]
        works_since = self.works_since
        works_until = self.works_until
        if works_since and version_num < works_since:
            return False
        elif works_until and works_until < version_num:
            return False
        else:
            return True

    def is_applicable_to(self, source_version: str, target_version: str) -> bool:
        return self.is_required_for(target_version) and self.is_compatible_with(source_version)


# NOTE (mb 2018-06-29): None of the fixers use asname. If a
#   module already has an import using asname, it won't be
#   detected as already imported, and another import (without the
#   asname) will be added to the module.
class ImportDecl(typ.NamedTuple):
    module_name    : str
    import_name    : typ.Optional[str]
    py2_module_name: typ.Optional[str]


# NOTE (mb 2018-06-24): This also includes builtins from py27
#   because a fixer might add a reference to it when building for
#   py27, in which case, we don't want other code to have
#   replaced it with something other than the old builtin.
#   For example replace range -> xrange, so it would be nice if
#   xrange isn't a string or something.
BUILTIN_NAMES = {
    "ArithmeticError",
    "AssertionError",
    "AttributeError",
    "BaseException",
    "BlockingIOError",
    "BrokenPipeError",
    "BufferError",
    "BytesWarning",
    "ChildProcessError",
    "ConnectionAbortedError",
    "ConnectionError",
    "ConnectionRefusedError",
    "ConnectionResetError",
    "DeprecationWarning",
    "EOFError",
    "Ellipsis",
    "EnvironmentError",
    "Exception",
    "False",
    "FileExistsError",
    "FileNotFoundError",
    "FloatingPointError",
    "FutureWarning",
    "GeneratorExit",
    "IOError",
    "ImportError",
    "ImportWarning",
    "IndentationError",
    "IndexError",
    "InterruptedError",
    "IsADirectoryError",
    "KeyError",
    "KeyboardInterrupt",
    "LookupError",
    "MemoryError",
    "ModuleNotFoundError",
    "NameError",
    "None",
    "NotADirectoryError",
    "NotImplemented",
    "NotImplementedError",
    "OSError",
    "OverflowError",
    "PendingDeprecationWarning",
    "PermissionError",
    "ProcessLookupError",
    "RecursionError",
    "ReferenceError",
    "ResourceWarning",
    "RuntimeError",
    "RuntimeWarning",
    "StopAsyncIteration",
    "StopIteration",
    "SyntaxError",
    "SyntaxWarning",
    "SystemError",
    "SystemExit",
    "TabError",
    "TimeoutError",
    "True",
    "TypeError",
    "UnboundLocalError",
    "UnicodeDecodeError",
    "UnicodeEncodeError",
    "UnicodeError",
    "UnicodeTranslateError",
    "UnicodeWarning",
    "UserWarning",
    "ValueError",
    "Warning",
    "ZeroDivisionError",
    "abs",
    "all",
    "any",
    "ascii",
    "bin",
    "bool",
    "bytearray",
    "bytes",
    "callable",
    "chr",
    "classmethod",
    "compile",
    "complex",
    "copyright",
    "credits",
    "delattr",
    "dict",
    "dir",
    "display",
    "divmod",
    "enumerate",
    "eval",
    "exec",
    "filter",
    "float",
    "format",
    "frozenset",
    "get_ipython",
    "getattr",
    "globals",
    "hasattr",
    "hash",
    "help",
    "hex",
    "id",
    "input",
    "int",
    "isinstance",
    "issubclass",
    "iter",
    "len",
    "license",
    "list",
    "locals",
    "map",
    "max",
    "memoryview",
    "min",
    "next",
    "object",
    "oct",
    "open",
    "ord",
    "pow",
    "print",
    "property",
    "range",
    "repr",
    "reversed",
    "round",
    "set",
    "setattr",
    "slice",
    "sorted",
    "staticmethod",
    "str",
    "sum",
    "super",
    "tuple",
    "type",
    "vars",
    "zip",
    "StandardError",
    "apply",
    "basestring",
    "buffer",
    "cmp",
    "coerce",
    "dreload",
    "execfile",
    "file",
    "intern",
    "long",
    "raw_input",
    "reduce",
    "reload",
    "unichr",
    "unicode",
    "xrange",
}

# In case somebody is working on py4k or something

BUILTIN_NAMES.update([name for name in dir(builtins) if not name.startswith("__")])
